File: Tests/kernels_if.c
#include "acc_testsuite.h"

int test(){
    int err = 0;
    srand(time(NULL));
    int data_on_device = 0;
    int * devtest = (int *)malloc(sizeof(int));
    real_t * a = (real_t *)malloc(n * sizeof(real_t));
    real_t * b = (real_t *)malloc(n * sizeof(real_t));
    devtest[0] = 1;

    for (int x = 0; x < n; ++x){
        a[x] = rand() / (real_t)(RAND_MAX / 10);
        b[x] = 0;
    }
    
    #pragma acc enter data copyin(devtest[0:1])
    #pragma acc parallel present(devtest[0:1])
    {
        devtest[0] = 0;
    }

    #pragma acc kernels if(data_on_device) present(a[0:n], b[0:n])
    {
        #pragma acc loop
        for (int x = 0; x < n; ++x){
            b[x] = a[x];
        }
    }

    for (int x = 0; x < n; ++x){
        if (fabs(b[x] - a[x]) > PRECISION){
            err += 1;
            printf("1\n");
        }
    }
   
    for (int x = 0; x < n; ++x){
        a[x] = rand() / (real_t)(RAND_MAX / 10);
        b[x] = 0;
    }

    #pragma acc enter data copyin(a[0:n]) create(b[0:n])
    data_on_device = 1;

    #pragma acc kernels if(data_on_device) present(a[0:n], b[0:n])
    {
        #pragma acc loop
        for (int x = 0; x < n; ++x){
            b[x] = a[x];
        }
    }
    #pragma acc exit data copyout(b[0:n]) delete(a[0:n])
    
    for (int x = 0; x < n; ++x){
        if (fabs(b[x] - a[x]) > PRECISION){
            err += 1;
            printf("2\n");
        }
    }

    if (devtest[0] == 1){
        for (int x = 0; x < n; ++x){
            a[x] = rand() / (real_t)(RAND_MAX / 10);
            b[x] = 0;
        }
        
        #pragma acc enter data copyin(a[0:n]) create(b[0:n])
        for (int x = 0; x < n; ++x){
            a[x] = -1;
        }

        #pragma acc kernels if(data_on_device) present(a[0:n], b[0:n])
        {
            #pragma acc loop
            for (int x = 0; x < n; ++x){
                b[x] = a[x];
            }
        }
        for (int x = 0; x < n; ++x){
            if (fabs(a[x] + 1) > PRECISION){
                err += 1;
                printf("3\n");
            }
            if (fabs(b[x]) > PRECISION){
                err += 1;
                printf("4\n");
            }
        }
        #pragma acc exit data copyout(a[0:n], b[0:n])
        data_on_device = 0;
        for (int x = 0; x < n; ++x){
            if (fabs(a[x] - b[x]) > PRECISION){
                err += 1;
                printf("5\n");
            }
        }

        for (int x = 0; x < n; ++x){
            a[x] = rand() / (real_t)(RAND_MAX / 10);
            b[x] = 0;
        }
        
        #pragma acc enter data copyin(a[0:n], b[0:n])
        #pragma acc kernels if(data_on_device) present(a[0:n], b[0:n])
        {
            #pragma acc loop
            for (int x = 0; x < n; ++x){
                b[x] = a[x];
            }
        }
        for (int x = 0; x < n; ++x){
            if (fabs(a[x] - b[x]) > PRECISION) {
                err += 1;
                printf("6\n");
            }
        }
        #pragma acc exit data copyout(a[0:n], b[0:n])
        for (int x = 0; x < n; ++x){
            if (fabs(b[x]) > PRECISION && b[x] != a[x]){
                err += 1;
                printf("7\n");
                printf("%f\n", b[x]);
            }
        }
    }


    free(a);
    free(b);
    return err;
}


int main()
{
  int i;			/* Loop index */
  int result;		/* return value of the program */
  int failed=0; 		/* Number of failed tests */
  int success=0;		/* number of succeeded tests */
  static FILE * logFile;	/* pointer onto the logfile */
  static const char * logFileName = "test_acc_lib_acc_wait.log";	/* name of the logfile */


  /* Open a new Logfile or overwrite the existing one. */
  logFile = fopen(logFileName,"w+");

  printf("######## OpenACC Validation Suite V %s #####\n", ACCTS_VERSION );
  printf("## Repetitions: %3d                       ####\n",REPETITIONS);
  printf("## Array Size : %.2f MB                 ####\n",ARRAYSIZE * ARRAYSIZE/1e6);
  printf("##############################################\n");
  printf("Testing test_acc_lib_acc_wait\n\n");

  fprintf(logFile,"######## OpenACC Validation Suite V %s #####\n", ACCTS_VERSION );
  fprintf(logFile,"## Repetitions: %3d                       ####\n",REPETITIONS);
  fprintf(logFile,"## Array Size : %.2f MB                 ####\n",ARRAYSIZE * ARRAYSIZE/1e6);
  fprintf(logFile,"##############################################\n");
  fprintf(logFile,"Testing test_acc_lib_acc_wait\n\n");

  for ( i = 0; i < REPETITIONS; i++ ) {
    fprintf (logFile, "\n\n%d. run of test_acc_lib_acc_wait out of %d\n\n",i+1,REPETITIONS);
    if (test() == 0) {
      fprintf(logFile,"Test successful.\n");
      success++;
    } else {
      fprintf(logFile,"Error: Test failed.\n");
      printf("Error: Test failed.\n");
      failed++;
    }
  }

  if(failed==0) {
    fprintf(logFile,"\nDirective worked without errors.\n");
    printf("Directive worked without errors.\n");
    result=0;
  } else {
    fprintf(logFile,"\nDirective failed the test %i times out of %i. %i were successful\n",failed,REPETITIONS,success);
    printf("Directive failed the test %i times out of %i.\n%i test(s) were successful\n",failed,REPETITIONS,success);
    result = (int) (((double) failed / (double) REPETITIONS ) * 100 );
  }
  printf ("Result: %i\n", result);
  return result;
}

File: Tests/kernels_if.f90
      INTEGER FUNCTION test()
        IMPLICIT NONE
        INCLUDE "acc_testsuite.f90"
        INTEGER :: x !Iterators
        REAL(8),DIMENSION(LOOPCOUNT):: a, b !Data
        INTEGER :: errors = 0
        LOGICAL,DIMENSION(1):: devtest
        LOGICAL:: data_on_device = .FALSE.
        devtest(1) = .TRUE.

        !$acc enter data copyin(devtest(1:1))
        !$acc kernels present(devtest(1:1))
          devtest(1) = .FALSE.
        !$acc end kernels

        !Initilization
        CALL RANDOM_SEED()
        CALL RANDOM_NUMBER(a)
        b = 0

        !$acc kernels if(data_on_device) present(a(1:LOOPCOUNT), b(1:LOOPCOUNT))
          !$acc loop
          DO x = 1, LOOPCOUNT
            b(x) = a(x)
          END DO
        !$acc end kernels

        DO x = 1, LOOPCOUNT
          IF (abs(a(x) - b(x)) .gt. PRECISION) THEN
            errors = errors + 1
          END IF
        END DO

        CALL RANDOM_NUMBER(a)
        b = 0

        data_on_device = .TRUE.
        !$acc enter data copyin(a(1:LOOPCOUNT)) create(b(1:LOOPCOUNT))
        !$acc kernels if(data_on_device) present(a(1:LOOPCOUNT), b(1:LOOPCOUNT))
          !$acc loop
          DO x = 1, LOOPCOUNT
            b(x) = a(x)
          END DO
        !$acc end kernels
        !$acc exit data copyout(b(1:LOOPCOUNT)) delete(a(1:LOOPCOUNT))
        
        DO x = 1, LOOPCOUNT
          IF (abs(b(x) - a(x)) .gt. PRECISION) THEN
            errors = errors + 1
          END IF
        END DO

        IF (devtest(1) .eqv. .TRUE.) THEN
          CALL RANDOM_NUMBER(a)
          b = 0
 
          !$acc enter data copyin(a(1:LOOPCOUNT)) create(b(1:LOOPCOUNT))
          DO x = 1, LOOPCOUNT
            a(x) = -1
          END DO
          
          !$acc kernels if(data_on_device) present(a(1:LOOPCOUNT), b(1:LOOPCOUNT))
            !$acc loop
            DO x = 1, LOOPCOUNT
              b(x) = a(x)
            END DO
          !$acc end kernels
 
          DO x = 1, LOOPCOUNT
            IF (abs(a(x) + 1) .gt. PRECISION) THEN
              errors = errors + 1
            END IF
            IF (abs(b(x)) .gt. PRECISION) THEN
              errors = errors + 1
            END IF
          END DO
 
          !$acc exit data copyout(a(1:LOOPCOUNT), b(1:LOOPCOUNT))
          data_on_device = .FALSE.

          DO x = 1, LOOPCOUNT
            IF (abs(a(x) - b(x)) .gt. PRECISION) THEN
              errors = errors + 1
            END IF
          END DO

          CALL RANDOM_NUMBER(a)
          b = 0

          !$acc enter data copyin(a(1:LOOPCOUNT), b(1:LOOPCOUNT))
          !$acc kernels if(data_on_device) present(a(1:LOOPCOUNT), b(1:LOOPCOUNT))
            !$acc loop
            DO x = 1, LOOPCOUNT
              b(x) = a(x)
            END DO
          !$acc end kernels

          DO x = 1, LOOPCOUNT
            IF (abs(a(x) - b(x)) .gt. PRECISION) THEN
              errors = errors + 1
            END IF
          END DO

          !$acc exit data copyout(a(1:LOOPCOUNT), b(1:LOOPCOUNT))
          
          DO x = 1, LOOPCOUNT
            IF (abs(b(x)) .gt. PRECISION) THEN
              IF (abs(b(x) - a(x)) .gt. PRECISION) THEN
                errors = errors + 1
              END IF
            END IF
          END DO
        END IF

        test = errors
      END


      PROGRAM test_kernels_async_main
      IMPLICIT NONE
      INTEGER :: failed, success !Number of failed/succeeded tests
      INTEGER :: num_tests,crosschecked, crossfailed, j
      INTEGER :: temp,temp1
      INCLUDE "acc_testsuite.f90"
      INTEGER test


      CHARACTER*50:: logfilename !Pointer to logfile
      INTEGER :: result

      num_tests = 0
      crosschecked = 0
      crossfailed = 0
      result = 1
      failed = 0

      !Open a new logfile or overwrite the existing one.
      logfilename = "OpenACC_testsuite.log"
!      WRITE (*,*) "Enter logFilename:"
!      READ  (*,*) logfilename

      OPEN (1, FILE = logfilename)

      WRITE (*,*) "######## OpenACC Validation Suite V 2.5 ######"
      WRITE (*,*) "## Repetitions:", N
      WRITE (*,*) "## Loop Count :", LOOPCOUNT
      WRITE (*,*) "##############################################"
      WRITE (*,*)

      WRITE (*,*) "--------------------------------------------------"
      WRITE (*,*) "Test of kernels_if"
      WRITE (*,*) "--------------------------------------------------"

      crossfailed=0
      result=1
      WRITE (1,*) "--------------------------------------------------"
      WRITE (1,*) "Test of kernels_if"
      WRITE (1,*) "--------------------------------------------------"
      WRITE (1,*)
      WRITE (1,*) "(Crosstests should fail)"
      WRITE (1,*)

      DO j = 1, N
        temp =  test()
        IF (temp .EQ. 0) THEN
          WRITE (1,*)  j, ". test successfull."
          success = success + 1
        ELSE
          WRITE (1,*) "Error: ",j, ". test failed."
          failed = failed + 1
        ENDIF
      END DO


      IF (failed .EQ. 0) THEN
        WRITE (1,*) "Directive worked without errors."
        WRITE (*,*) "Directive worked without errors."
        result = 0
        WRITE (*,*) "Result:",result
      ELSE
        WRITE (1,*) "Directive failed the test ", failed, " times."
        WRITE (*,*) "Directive failed the test ", failed, " times."
        result = failed * 100 / N
        WRITE (*,*) "Result:",result
      ENDIF
      CALL EXIT (result)
      END PROGRAM
                                             

